Group by category changes

kingy04 opened the pull request:
This brings in changes to enable us to group events by category's by passing a group_by prop on events and passing in a new categoryGrouping array to specify an ordered list of category's or a boolean
- True: to generate the category list from the values in the group_by prop
- False: To place all events in a single misc category

It also adds a scoped slot for passing in a custom category header, adds tree-shaking for our use of lodash and utilises an iso caret icon for the default category headers to show the collapse state

and any logic/styling changes to go with the above, it also adds the semantic ui css for use when testing the repo locally

A seperate PR for tests for the new groupByData logic will follow this in the near future.

kingy04 commented:
These changes can be tested on the forward planning gantt on supercroud dashboards on this branch feature/CC-1176-Gantt-chart-category-changes

BrockReece commented:
There seems to be an vertical offset issue as the rows aren't matching up, I guess is down to how the box model is calculating the height. The new ToggleHeader.vue has a fixed height of 35px and a border of 1px which gives it a total height of 37px, 2px larger than the spacer rectangle on the other side of the graph. 

It's subtle enough for a small number of rows but could compound to a major offset with a real dataset

kingy04 commented:
Thanks for the review, i have made the follow changes:

- Misc header is not shown if category groupings is set to false
- Show misc at the bottom when dynamically categorised (Im not sure we stated that misc would be weighed at the bottom when dynamically grouping the categories, although when passing our array of categories misc would have been at the last index (as this was mentioned).. but its logical/consistent so i have set it so it will show at the bottom in both cases)
- Fix vertical offset, i had fixed one instance of this but it reoccured when i moved the border style responsibility to the gantt (Good spot)

BrockReece commented:
Ok, I will review again over the weekend. Thanks

BrockReece commented:
Sorry for the ambiguity.

As in if we pass in `:categoryGroupings="false"`, we could have treated this as a single category grouping. Then we wouldn't have to maintain two almost identical sets of template markup and node building logic. 

The logic I built out (specifically line 447) already does this by effectively putting everything into the `misc` category if the categoryGroupings prop is `false`.

paul-watkins-croud commented:
@BrockReece @kingy04 Just to add absolutely no value but to clarify, this was the logic that we all agreed on collectively before respective holidays wasn't it? I'd like to understand where the communication breakdown has arisen from to avoid this kind of issue again.

kingy04 commented on src/components/Gantt.vue:
```diff
@@ -2,7 +2,27 @@
         <div id="adpcalendar">
             <div id="timeline-sidebar">
-                <svg id="event-types" ref="svg" :width="titleWidth" :height='limits.height'>
+                <div v-if="categoryGroupings" v-for="(i, index) in groupByData" :key="index" :style="{height: `${i.show ? i.height + getTopMargin : getTopMargin}px`}">
```
Yes we can, that is better. Done

kingy04 commented on src/components/Gantt.vue:
```diff
@@ -14,60 +34,91 @@
-                                            <title>{{ child.title }}</title>
-                                        </rect>
+                                    <g class="blocks">
+                                        <g v-if="categoryGroupings" class="block" v-for="(block, $index) in i.blocks" :key="$index">
```
done

kingy04 commented on src/components/Gantt.vue:
```diff
@@ -373,6 +434,54 @@
+                    position(item)
+
+                    const filteredGroups = {}
+                    Object.keys(grouped).forEach((prop) => {
```
Your right, i have moved it outside the loop

kingy04 commented on src/components/Gantt.vue:
```diff
@@ -353,9 +410,13 @@
-                    marginTop: '20px',
+                    marginTop: `${this.getTopMargin}px`,
                     marginLeft: `${this.titleWidth}px`,
                     height: `${this.limits.height}px`,
```
good spot

BrockReece commented:
@kingy04 No, my point was that we weren't doing that in this PR, you were using different node building and template logic based on the `categoryGroupings` prop. Cleaning up the code is something we should be doing anyway.

Yes, we did discuss this and yes it is what was implemented in the logic that I built before I went away, but that is where this idea was left behind it seems.

As for all the new issues found that breaks existing functionality, this is a nod to the comment I made about building more tech debt into our projects and this is what I understood was being worked on since I left for my holiday?

BrockReece commented:
@kingy04 I think there is an issue with reactivity, the nodes aren't rebuilding if there is a change in the data, this is probably causing the 'postpone' issue you mentioned above, but it is also causing 'children' blocks not to update when you drag and drop the parent block.

BrockReece commented on src/components/Gantt.vue:
```diff
@@ -373,9 +370,77 @@
+
+                    if (item.dependencies) {
+                        item.dependencies.map((dep) => {
+                            this.links.push([
```
We could treat these links like everything else in `buildGroupByData`  and scope it to a category, which would fix links being associated to the wrong blocks and bleeding into other categories as seen below

<img width="1428" alt="screen shot 2018-05-14 at 10 59 59" src="https://user-images.githubusercontent.com/8988409/39991009-0e6e356a-5766-11e8-9c6f-b3497e1fe893.png">

BrockReece commented:
Just fixed some `:key` based errors to get rid of some noise in the console logs

kingy04 commented:
I have added a watcher to update the event positions, do you think that is ok or is there a better way to deal with it?, we want to avoid re building the data to often right?

Ok ill have a look at the link logic

BrockReece commented:
I think that will only catch a very narrow set of data updates. This only considers the offset of the event that is currently selected in the ui, right?

I don't see an issue with re running the `groupByData` when the underlying data changes, the most expensive part will probably be updating the DOM and because of Vue's Virtual DOM, this will only re render the branches with actual changes.

If we are concerned with performance, we could abstract the node building logic out of the groupByData method and into it's own `buildEvent` method. We could then put a watcher on all the event data, compare the diff of the change and only run the rebuild for the specific events. But I don't think this will give us much of a return on the investment.

BrockReece commented:
I think a good rule of thumb is that data integrity should always be our primary concern and performance improvements should only be considered after that.

snipey101 commented:
This has passed product QA, so I'm happy

On Mon, 18 Jun 2018, 11:24 Brock Reece, <notifications@github.com> wrote:

> @kingy04 <https://github.com/kingy04> @paul-watkins-croud
> <https://github.com/paul-watkins-croud>
>
> Can we confirm that we are confident and happy for this to be merged in to
> master and I will approve and bump the major version number.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CroudTech/croud-vue-gantt/pull/13#issuecomment-398010031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAysdR8chrm7ZRnFXX7-ZN2cyCuCxx2mks5t93_BgaJpZM4T50dI>
> .
>

kingy04 commented:
@BrockReece, Im happy for this to be merged to master. Thanks

BrockReece commented:
This has been released in v2.0.0
